### ISSUE TITLE
Signup form asking for username despite local_iomad_signup_usemail being set to true

### ISSUE THREAD

**Issue by ThomE2** [iomad/iomad#1613]
When a new user comes to the site, he has no section active and then no company is set in the file ``login/signup.php`` on line 39. 
Then ``if (!empty($this->company))`` on line 44 return false, and the condition on line 45  for local_iomad_signup_useemail is never checked. Then the form displays its "normal" configuration, asking for a username instead of asking for the email address and saying that this address will be used as username. 

This really seems weird to me because most of the time, if a user wants to signup, he is most likely to NOT have a session yet. However we would like to create new users with email as username in all cases. 

Is there any way to make it so ? Why this ``if (!empty($this->company))`` in the first place ? The fact that we want to use the email is a global setting which does not depend on the company ...

Thank for your help !

**Comment by thomaseudes** [comment]
Furthermore, when there is no session company, and the form is in its "username state" (and not in its "email address one"), trying to sign up results in this error : ``Exception : Call to a member function get_theme() on bool``

```Stack trace

    line 104 of /local/iomad_signup/lib.php: Error thrown
    line 39 of /local/iomad_signup/classes/observer.php: call to local_iomad_signup_user_created()
    line ? of unknownfile: call to local_iomad_signup_observer::user_created()
    line 155 of /lib/classes/event/manager.php: call to call_user_func()
    line 75 of /lib/classes/event/manager.php: call to core\event\manager::process_buffers()
    line 835 of /lib/classes/event/base.php: call to core\event\manager::dispatch()
    line 157 of /auth/email/auth.php: call to core\event\base->trigger()
    line 99 of /auth/email/auth.php: call to auth_plugin_email->user_signup_with_confirmation()
    line 159 of /login/signup.php: call to auth_plugin_email->user_signup()
    ```

**Comment by thomaseudes** [comment]
(When creating a user while the form is its "username state" rather than its "email state", the user is created anyway with the email address as its username. This might be very confusing for new users who enter a username and then cannot access their account with it)

**Comment by ThomE2** [comment]
Furthermore, the users created this way do have their email address as username, even if they saw the form asking for a username. This might be very confusing for a new user to enter a username and then not be able to access their account with it

**Comment by ThomE2** [comment]
I also noted that if there is an active session, a user created through the email self-registration will automatically be added to the company of that session, rather than the company linked to his email domain or the default company defined in the settings. 
Is that the wanted behavior ? In my opinion, it doesn't make much sense, let's imagine a user having an account in company A, who disconnects from this account, and then starts creating an account for an other user member of company B, the second user will then be added to company A even if his email address is something@company-b.com

**Comment by turf212** [comment]
The original issue was fixed in commit 642d554fbf87.

Is it likely that a user will sign into company A and then log out and the try and create a new account with company B?  I can see that if they were the same user in two different companies but to them (without URL switching) it would be the same site.  If you were using URL switching/per company URLs then they would go to the different URL to create an account.

Keeping the company information on the session allows for branded log out/log in pages.  There is no other way of doing this as without it how do you know what company a user is in _before_  they log in?

**Comment by turf212** [comment]
On the company matching - URL parameters trump everything else.  The domain matching is secondary then the default company is last.  If your URL matches a company then that is the company you will be assigned to as it avoids issues like outlook.com and so on email addresses.
